Gems/PrefabDependencyViewer/Code/Source/DirectedGraph.h

```diff
@@ -0,0 +1,231 @@
+                stack.push(AZStd::make_pair(rhs.m_root, nullptr));
+
+                while (!stack.empty())
+                {
+                    AZStd::pair<Node*, Node*> pair = stack.top();
+                    Node* rhsNode = pair.first;
+                    Node* parent = pair.second;
+
+                    stack.pop();
+
+                    MetaData cpyMetaData = rhsNode->GetMetaData();
+                    Node* copy = aznew Node(cpyMetaData.GetTemplateId(),
+                                            cpyMetaData.GetSource(),
+                                            parent);
+
+                    AddNode(copy);
+                    AddChild(parent, copy);
+
+                    for (Node* child : rhs.GetChildren(rhsNode))
+                    {
+                        stack.push(AZStd::make_pair(child, copy));
+                    }
+                }
+            }
+
+            ~DirectedGraph()
+            {
+                auto delete_node = [](Node* node) { delete node; };
+                AZStd::for_each(m_nodes.begin(), m_nodes.end(), delete_node);
+            }
+
+            AZStd::vector<int> countNodesAtEachLevel(int& widestLevelSize) const
+            {
+                /** Directed Graph can't have cycles because of the
+                non-circular nature of the Prefab. */
+                AZStd::vector<int> count;
+                using pair = AZStd::pair<int, Node*>;
+                AZStd::queue<pair> queue;
+                queue.push(AZStd::make_pair(0, m_root));
+
+
+                while (!queue.empty())
+                {
+                    pair p = queue.front();
+                    int level = p.first;
+                    Node* currNode = p.second;
+
+                    queue.pop();
+
+                    if (count.size() <= level)
+                    {
+                        // Started a new level so check if the previous level
+                        // was any bigger then the widest level size seen so far.
+                        int prevLevelCount = level != 0 ? count[level - 1] : 0;
+                        widestLevelSize = AZStd::max(widestLevelSize, prevLevelCount);
+                        count.push_back(1);
+                    }
+                    else
+                    {
+                        ++count[level];
+                    }
+
+                    auto it = m_children.find(currNode);
+                    if (it != m_children.end())
+                    {
+                        NodeSet children = it->second;
+                        for (Node* node : children)
+                            queue.push(AZStd::make_pair(level + 1, node));
+                    }
+                }
+
+                // Check if the last level was the widest.
+                widestLevelSize = AZStd::max(widestLevelSize, count[count.size() - 1]);
+                return count;
+            }
+        private:
+            NodeSet m_nodes;
```

Comment:
We should make the `m_nodes` set own the nodes, by changing it to be a `AZStd::unordered_set<AZStd::unique_ptr<Node>>`
```suggestion
            AZStd::unordered_set<AZStd::unique_ptr<Node>> m_nodes;
```

With the current design, we still need some `NodeSet` type that doesn't own the nodes in it.

Reply:
Do we get any benefit out of having one giant set that contains all the nodes? If not, I think it would be cleaner to restructure the Node class to own its children. Then you'd have:
```cpp
class Node
{
    AZStd::unordered_set<AZStd::unique_ptr<Node>> m_children;
    Node* m_parent = nullptr;
};
class DirectedGraph
{
    AZStd::unique_ptr<Node> m_root;
};
```

Reply:
Can't we hold onto this change and get some more feedback and do it unless we absolutely have too? There is more work to be done on the Viewer which can take some time. I'm not saying that we should compromise on code quality but I think at this moment priority should be to get the tool fully functional.